sympy/diffgeom/diffgeom.py

```diff
@@ -1084,6 +1086,390 @@ def __call__(self, field):
     def _latex(self, printer, *args):
         return r'\mathbb{\nabla}_{%s}' % printer._print(self._wrt)
 
+##########################################################################
+# Tensors as multidimensional arrays.
+##########################################################################
+
+class TensorArray(object):
+    """
+    Object representing a tensor as a multidimensional sparse symbolic array.
+
+    ``TensorArray(components=..., variance=..., coordinate_system=...)``:
+    initializes a tensor array from a given component array, with given
+    variance and coordinate_system (see ``CoordSystem``).  The variance should
+    be a list of +1 or -1.  Where +1 slots are covariant and -1 slots are
+    contravariant.
+
+    For conversion between differential geometry objects and ``TensorArray``s,
+    see ``ToArray``.
+
+    Example: Converting the Riemann components (an array) to a proper tensor
+    ========================================================================
+
+    >>> from sympy.diffgeom import CoordSystem, Patch, TensorArray, TensorProduct, Manifold, metric_to_Riemann_components
+    >>> from sympy import Function
+    >>> from sympy.functions import sin
+    >>> M = Manifold('Reisner-Nordstrom', 4)
+    >>> p = Patch('origin', M)
+    >>> cs = CoordSystem('spherical', p, ['t', 'r', 'theta', 'phi'])
+    >>> t, r, theta, phi = cs.coord_functions()
+    >>> dt, dr, dtheta, dphi = cs.base_oneforms()
+    >>> f=Function('f')
+    >>> TP=TensorProduct
+    >>> metric = f(r)**2*TP(dt, dt) - f(r)**(-2)*TP(dr, dr) - r**2*TP(dtheta, dtheta) - r**2*sin(theta)**2*TP(dphi, dphi)
+    >>> rm = TensorArray(components=metric_to_Riemann_components(metric), variance=[-1,1,1,1],coordinate_system=cs)  #Note the variance!
+
+    """
+    def _contravariant_slots(self,T):
+        slots=[]
+        for k in range(self.order):
+            l = [None for i in range(self.order-1)]
+            l.insert(k,self.base_vectors[0])
+            try:
+                T.rcall(*l)
+            except:
+                slots.append(k)
+        return slots
+
+    def __init__(self, **kwargs):
```

Comment:
Please use named parameters instead of keywords. Then the compiler will check that the expected parameters are given. That will also be helpful to the user who runs `help(TensorArray)`.

Reply:
Ok